src/ConnectionManager.ts

```diff
@@ -91,12 +91,56 @@ export class ConnectionManager extends EventEmitter {
         throw new ApiError(`Connection type not known`);
     }
 
-    private assertStateAllowed(state: StateEvent<any>, serviceType: string) {
-        if (state.sender === this.as.botUserId) {
-            return;
+    /**
+     * Check if a state event is sent by a user who is allowed to configure the type of connection the state event covers.
+     * If it isn't, revert the state to the last-known valid value, or redact it if that isn't possible.
+     * @param roomId The target Matrix room.
+     * @param state The state event for altering a connection in the room.
+     * @param serviceType The type of connection the state event is altering.
+     * @returns Whether the state event was allowed to be set. If not, the state will be reverted asynchronously.
+     */
+    public verifyStateEvent(roomId: string, state: StateEvent, serviceType: string) {
+        if (!this.isStateAllowed(roomId, state, serviceType)) {
+            void this.tryRestoreState(roomId, state, serviceType);
+            log.error(`User ${state.sender} is disallowed to manage state for ${serviceType} in ${roomId}`);
+            return false;
+        } else {
+            return true;
         }
-        if (!this.config.checkPermission(state.sender, serviceType, BridgePermissionLevel.manageConnections)) {
-            throw new Error(`User ${state.sender} is disallowed to create state for ${serviceType}`);
+    }
+
+    /**
+     * The same as {@link verifyStateEvent}, but verifies the state event against the room & service type of the given connection.
+     * @param connection The connection to verify the state event against.
+     * @param state The state event for altering a connection in the room targeted by {@link connection}.
+     * @returns Whether the state event was allowed to be set. If not, the state will be reverted asynchronously.
+     */
+    public verifyStateEventForConnection(connection: IConnection, state: StateEvent) {
+        const cd: ConnectionDeclaration = Object.getPrototypeOf(connection).constructor;
+        return !this.verifyStateEvent(connection.roomId, state, cd.ServiceCategory);
+    }
+
+    private isStateAllowed(roomId: string, state: StateEvent, serviceType: string) {
+        return state.sender === this.as.botUserId
+            || this.config.checkPermission(state.sender, serviceType, BridgePermissionLevel.manageConnections);
+    }
+
+    private async tryRestoreState(roomId: string, originalState: StateEvent, serviceType: string) {
+        let state = originalState;
+        let attemptsRemaining = 5;
+        try {
+            do {
+                if (state.unsigned.replaces_state) {
+                    state = new StateEvent(await this.as.botClient.getEvent(roomId, state.unsigned.replaces_state));
+                } else {
+                    await this.as.botClient.redactEvent(roomId, originalState.eventId,
```

Comment:
I have worries this might cause a problem *if* you misconfigure and restart your hookshot, maybe this should only apply after restart?

Reply:
Sounds reasonable.  Then how should a disallowed state event be handled on startup?  Since the goal is to guard against a bad config redacting valid state events, should Hookshot just fail to start?

Reply:
Disallowed -> Ignore the connection entirely.

Reply:
This is now addressed in 1bc4c25 by not rolling back invalid state events on startup.